src/transformers/models/ibert/utils/quant_utils.py

```diff
@@ -0,0 +1,264 @@
+        # lower_index += 1
+    else:
+        lower_bound = -torch.kthvalue(-input, k=lower_index).values
+
+    if not output_tensor:
+        lower_bound = lower_bound.item()
+        upper_bound = upper_bound.item()
+    return lower_bound, upper_bound
+
+
+def linear_quantize(input, scale, zero_point, inplace=False):
+    """
+    Quantize single-precision input tensor to integers with the given scaling factor and zeropoint.
+
+    Parameters:
+    ----------
+    input: single-precision input tensor to be quantized
+    scale: scaling factor for quantization
+    zero_pint: shift for quantization
+    """
+
+    # reshape scale and zeropoint for convolutional weights and activation
+    if len(input.shape) == 4:
+        scale = scale.view(-1, 1, 1, 1)
+        zero_point = zero_point.view(-1, 1, 1, 1)
+    # reshape scale and zeropoint for linear weights
+    elif len(input.shape) == 2:
+        scale = scale.view(-1, 1)
+        zero_point = zero_point.view(-1, 1)
+    else:
+        scale = scale.view(-1)
+        zero_point = zero_point.view(-1)
+    # quantized = float / scale + zero_point
+    if inplace:
+        input.mul_(1. / scale).add_(zero_point).round_()
+        return input
+    return torch.round(1. / scale * input + zero_point)
+
+def symmetric_linear_quantization_params(num_bits,
+                                        saturation_min,
+                                        saturation_max,
+                                        per_channel=False):
+    """
+    Compute the scaling factor with the given quantization range for symmetric quantization.
+
+    Parameters:
+    ----------
+    saturation_min: lower bound for quantization range
+    saturation_max: upper bound for quantization range
+    """
+    # in this part, we do not need any gradient computation,
+    # in order to enfore this, we put torch.no_grad()
+    with torch.no_grad():
+        n = 2 ** (num_bits - 1) - 1
+
+        if per_channel:
+            scale, _ = torch.max(torch.stack([saturation_min.abs(), saturation_max.abs()], dim=1), dim=1)
+            scale = torch.clamp(scale, min=1e-8) / n 
+
+        else:
+            scale = max(saturation_min.abs(), saturation_max.abs())
+            scale = torch.clamp(scale, min=1e-8) / n 
+
+    return scale 
+
+
+class SymmetricQuantFunction(Function):
+    """
+    Class to quantize the given floating-point values using symmetric quantization with given range and bitwidth.
+    """
+    @staticmethod
+    def forward(ctx, x, k, percentile_mode=False, specified_scale=None):
+        """
+        x: floating point tensor to be quantized
+        k: quantization bitwidth
+        Note that the current implementation of SymmetricQuantFunction requires pre-calculated scaling factor.
+        specified_scale: pre-calculated scaling factor for the tensor x
+        """
+        
+        if specified_scale is not None:
+            scale = specified_scale
+
+        zero_point = torch.tensor(0.).cuda()
```

Comment:
Does this only work on GPU? -> Is is not possible to run those functions on CPU?